### ISSUE TITLE
USWDS - Bug: Step indicator - 508 issues

### ISSUE THREAD

**Issue by jenstrickland** [uswds/uswds#5294]
### Describe the bug

# USWDS - Bug: Step indicator - 508 issues

## Details

USWDS component: Step Indicator https://designsystem.digital.gov/components/step-indicator/

### Issues

In Safari the ordered list will not be read as a list because the list display is over-ridden in the CSS with `list-style:none`.
 

### Potential Fix

To fix this, add `role="list"` to the `<ol>` and `role="list item` to the `<li>`. 

## Additional Issue

Another issue noted by a colleague in an accessibility audit in 2021: 
> "the visual organization of content and the reading order in the component preview is confusing: when using the arrow down, the heading identifying the component is announced after Step 3 of 5 Supporting documents. For example, JAWS reads: Supporting documents Step 3 of 5 and then No Labels"

### Steps to reproduce the bug

1. Go to https://designsystem.digital.gov/components/step-indicator/
2. View component preview, component code, and inspect the component preview code to view the CSS that removes default list styling using `list-style: none`
3. Use VoiceOver to read the component in Safari

### Expected Behavior

- 1.3.1 Info and Relationships: Information, [structure](https://www.w3.org/TR/WCAG20/#structuredef), and [relationships](https://www.w3.org/TR/WCAG20/#relationshipsdef) conveyed through [presentation](https://www.w3.org/TR/WCAG20/#presentationdef) can be [programmatically determined](https://www.w3.org/TR/WCAG20/#programmaticallydetermineddef) or are available in text. (Level A)
- 1.3.2 Meaningful Sequence: When the sequence in which content is presented affects its meaning, a [correct reading sequence](https://www.w3.org/TR/WCAG20/#correct-reading-sequencedef) can be [programmatically determined](https://www.w3.org/TR/WCAG20/#programmaticallydetermineddef). (Level A)

### Current Code

```html
<h3 class="site-preview-heading">Default</h3>
<div class="usa-step-indicator" aria-label="progress">
  <ol class="usa-step-indicator__segments">
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Personal information <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Household status <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--current"
      aria-current="true"
    >
      <span class="usa-step-indicator__segment-label"
        >Supporting documents</span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Signature <span class="usa-sr-only">not completed</span></span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Review and submit <span class="usa-sr-only">not completed</span></span
      >
    </li>
  </ol>
  <div class="usa-step-indicator__header">
    <h4 class="usa-step-indicator__heading">
      <span class="usa-step-indicator__heading-counter"
        ><span class="usa-sr-only">Step</span>
        <span class="usa-step-indicator__current-step">3</span>
        <span class="usa-step-indicator__total-steps">of 5</span> </span
      ><span class="usa-step-indicator__heading-text"
        >Supporting documents</span
      >
    </h4>
  </div>
</div>

<h3 class="site-preview-heading">No labels</h3>
<div
  class="usa-step-indicator usa-step-indicator--no-labels"
  aria-label="progress"
>
  <ol class="usa-step-indicator__segments">
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Personal information <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Household status <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--current"
    >
      <span class="usa-step-indicator__segment-label"
        >Supporting documents</span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Signature <span class="usa-sr-only">not completed</span></span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Review and submit <span class="usa-sr-only">not completed</span></span
      >
    </li>
  </ol>
  <div class="usa-step-indicator__header">
    <h4 class="usa-step-indicator__heading">
      <span class="usa-step-indicator__heading-counter"
        ><span class="usa-sr-only">Step</span>
        <span class="usa-step-indicator__current-step">3</span>
        <span class="usa-step-indicator__total-steps">of 5</span> </span
      ><span class="usa-step-indicator__heading-text"
        >Supporting documents</span
      >
    </h4>
  </div>
</div>

<h3 class="site-preview-heading">Centered</h3>
<div
  class="usa-step-indicator usa-step-indicator--center"
  aria-label="progress"
>
  <ol class="usa-step-indicator__segments">
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Personal information <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Household status <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--current"
      aria-current="true"
    >
      <span class="usa-step-indicator__segment-label"
        >Supporting documents</span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Signature <span class="usa-sr-only">not completed</span></span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Review and submit <span class="usa-sr-only">not completed</span></span
      >
    </li>
  </ol>
  <div class="usa-step-indicator__header">
    <h4 class="usa-step-indicator__heading">
      <span class="usa-step-indicator__heading-counter"
        ><span class="usa-sr-only">Step</span>
        <span class="usa-step-indicator__current-step">3</span>
        <span class="usa-step-indicator__total-steps">of 5</span> </span
      ><span class="usa-step-indicator__heading-text"
        >Supporting documents</span
      >
    </h4>
  </div>
</div>

<h3 class="site-preview-heading">Counters</h3>
<div
  class="usa-step-indicator usa-step-indicator--counters"
  aria-label="progress"
>
  <ol class="usa-step-indicator__segments">
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Personal information <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Household status <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--current"
      aria-current="true"
    >
      <span class="usa-step-indicator__segment-label"
        >Supporting documents</span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Signature <span class="usa-sr-only">not completed</span></span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Review and submit <span class="usa-sr-only">not completed</span></span
      >
    </li>
  </ol>
  <div class="usa-step-indicator__header">
    <h4 class="usa-step-indicator__heading">
      <span class="usa-step-indicator__heading-counter"
        ><span class="usa-sr-only">Step</span>
        <span class="usa-step-indicator__current-step">3</span>
        <span class="usa-step-indicator__total-steps">of 5</span> </span
      ><span class="usa-step-indicator__heading-text"
        >Supporting documents</span
      >
    </h4>
  </div>
</div>

<h3 class="site-preview-heading">Small counters</h3>
<div
  class="usa-step-indicator usa-step-indicator--counters-sm"
  aria-label="progress"
>
  <ol class="usa-step-indicator__segments">
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Personal information <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Household status <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li
      class="usa-step-indicator__segment usa-step-indicator__segment--current"
      aria-current="true"
    >
      <span class="usa-step-indicator__segment-label"
        >Supporting documents</span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Signature <span class="usa-sr-only">not completed</span></span
      >
    </li>
    <li class="usa-step-indicator__segment">
      <span class="usa-step-indicator__segment-label"
        >Review and submit <span class="usa-sr-only">not completed</span></span
      >
    </li>
  </ol>
  <div class="usa-step-indicator__header">
    <h4 class="usa-step-indicator__heading">
      <span class="usa-step-indicator__heading-counter"
        ><span class="usa-sr-only">Step</span>
        <span class="usa-step-indicator__current-step">3</span>
        <span class="usa-step-indicator__total-steps">of 5</span> </span
      ><span class="usa-step-indicator__heading-text"
        >Supporting documents</span
      >
    </h4>
  </div>
</div>
``` 

### Screenshots

![20230517-StepIndicatorSafariList](https://github.com/uswds/uswds/assets/57469/9c12d347-3687-4b6e-a843-693782c83f78)

### System setup

- USWDS version: https://designsystem.digital.gov/components/step-indicator/
- Device: MacBook Pro
- Operating System: Mac
- Browser and version: Safari version 16.4

### Potential fix example

```html
<h3 class="site-preview-heading">Default</h3>
<div class="usa-step-indicator" aria-label="progress">
  <ol role="list" class="usa-step-indicator__segments">
    <li role="listitem"
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
      <span class="usa-step-indicator__segment-label"
        >Personal information <span class="usa-sr-only">completed</span></span
      >
    </li>
    <li role="listitem"
      class="usa-step-indicator__segment usa-step-indicator__segment--complete"
    >
 <!-- etcetera -->
``` 

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [X] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Bug%22) for duplicate bug reports.

**Comment by amyleadem** [comment]
Confirming this issue exists in Safari:
- Confirmed that VoiceOver reads the list items, but not list semantics in Safari
    - Both Chrome or Firefox read the list items and list semantics
    - VoiceOver Gestures in iOS does not read the list items at all, but that is expected in the simplified mobile view
- Confirmed that adding `role="list"` adds list semantics (“List 5 items”,  “1 of 5”, etc) in Safari

**Comment by danbrady** [comment]
Hi @amyleadem @jenstrickland! I restructured the elements for better screen reading order and added `role="list"` to fix the Safari/VoiceOver issue. Could you please validate the updates in #5345 when you have time? Thanks!

**Comment by jenstrickland** [comment]
> Hi @amyleadem @jenstrickland! I restructured the elements for better screen reading order and added `role="list"` to fix the Safari/VoiceOver issue. Could you please validate the updates in #5345 when you have time? Thanks!

Hi Dan!

Did you also add the `role="listitem"` to the `li` elements? 

Sorry, I'm on my phone after just seeing the comments while en route.

**Comment by danbrady** [comment]
Hey @jenstrickland! I updated the PR to include `role="listitem"` on the `li` elements. Please validate when you have time. Thanks!
